DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h

```diff
@@ -1,244 +1,286 @@
+                    ap_uint<TrackBitWidths::kChi2RZSize> chi2RZ,
+                    ap_uint<TrackBitWidths::kBendChi2Size> bendChi2,
+                    ap_uint<TrackBitWidths::kHitPatternSize> hitPattern,
+                    ap_uint<TrackBitWidths::kMVAQualitySize> mvaQuality,
+                    ap_uint<TrackBitWidths::kMVAOtherSize> mvaOther);
 
-  unsigned int get_hitPattern();
-  unsigned int get_ispare();  // will need a converter(s) when spare bits are defined
-
-  unsigned int get_RinvBits();
-  unsigned int get_phiBits();
-  unsigned int get_tanlBits();
-  unsigned int get_z0Bits();
-  unsigned int get_d0Bits();
-  unsigned int get_chi2XYBits();
-  unsigned int get_chi2ZBits();
-  unsigned int get_BendChi2Bits();
-
-  // copy constructor
-
-  TTTrack_TrackWord(const TTTrack_TrackWord& word) {
-    initialize();
-
-    iRinv = word.iRinv;
-    iphi = word.iphi;
-    itanl = word.itanl;
-    iz0 = word.iz0;
-    id0 = word.id0;
-    ichi2XY = word.ichi2XY;
-    ichi2Z = word.ichi2Z;
-    iBendChi2 = word.iBendChi2;
-    ispare = word.ispare;
-    iHitPattern = word.iHitPattern;
-
-    // three 32-bit packed words
-    TrackWord1 = word.TrackWord1;
-    TrackWord2 = word.TrackWord2;
-    TrackWord3 = word.TrackWord3;
+private:
+  // ----------private member functions --------------
+  unsigned int digitizeSignedValue(double value, unsigned int nBits, double lsb) const {
+    unsigned int digitized_value = std::floor(std::abs(value) / lsb);
+    unsigned int digitized_maximum = (1 << (nBits - 1)) - 1;  // The remove 1 bit from nBits to account for the sign
+    if (digitized_value > digitized_maximum)
+      digitized_value = digitized_maximum;
+    if (value < 0)
+      digitized_value = (1 << nBits) - digitized_value;  // two's complement encoding
+    return digitized_value;
   }
 
-  TTTrack_TrackWord& operator=(const TTTrack_TrackWord& word) {
-    initialize();
-    iRinv = word.iRinv;
-    iphi = word.iphi;
-    itanl = word.itanl;
-    iz0 = word.iz0;
-    id0 = word.id0;
-    ichi2XY = word.ichi2XY;
-    ichi2Z = word.ichi2Z;
-    iBendChi2 = word.iBendChi2;
-    ispare = word.ispare;
-    iHitPattern = word.iHitPattern;
-
-    // three 32-bit packed words
-    TrackWord1 = word.TrackWord1;
-    TrackWord2 = word.TrackWord2;
-    TrackWord3 = word.TrackWord3;
+  template <typename T>
+  constexpr unsigned int getBin(double value, const T& bins) const {
+    auto up = std::upper_bound(bins.begin(), bins.end(), value);
+    return (up - bins.begin() - 1);
+  }
 
-    return *this;
+  double unpackSignedValue(unsigned int bits, unsigned int nBits, double lsb) const {
+    int isign = 1;
+    unsigned int digitized_maximum = (1 << nBits) - 1;
+    if (bits & (1 << (nBits - 1))) {  // check the sign
+      isign = -1;
+      bits = (1 << (nBits + 1)) - bits;  // if negative, flip everything for two's complement encoding
+    }
+    return (double(bits & digitized_maximum) + 0.5) * lsb * isign;
```

Comment:
Thomas Schuh is an expert on this (but not a member of this git repo). Representing the digi, signed quantity by an integer J and the float by F, the usual translation is:
1) F = (J+0.5)*lsb
2) J = floor(F/lsb)
so with lsb=1: J=-2,-1,0,1 become F=-1.5,-0.5,0.5,1.5 . There is no missing value at -0.5.
If F=-2.5, I'd expect J=-3, and converting this to a 3 bit  2s-complement no. gives 0b101 = 5 https://en.wikipedia.org/wiki/Two%27s_complement . This seems inconsistent with the maths in digitizeSignedValue(), which for value=-2.5 & nBits=3 would give 2s complement output of (8 - 2)=6?  And unpackedSignedValue() looks odd too.

Reply:
Hi @tomalin, 
Well, except for possibly a few corner cases, round(x) is equal to floor(x+0.5), and so the difference between my proposal `F = J*lsb   ;   J = round(F/lsb)`  and your `F = (J+0.5)*lsb   ;   J = floor(F/lsb)`  is just a shift of 0.5 LSB in the translation.
I still prefer mine, since it keeps the integer zero centered to the floating point zero.

Reply:
Hi @gpetruc,
I distinguish between biased integers `J = round(F / lsb); F = J * lsb;` and unbiased integers `J = floor(F / lsb); F = (J + .5) * lsb;`. If this TTTrack_TrackWord is supposed to be equal to the word being used in the h/w, one likely must use the same kind of integers as the h/w, which are biased integers.

Reply:
the naming seems a bit confusing (I would have called the `F = (J+.5)*LSB` the biased one...) but anyway yes, this class should use the same encoding that the firmware will use so that we would not need to change the decoding when processing the output of the hardware (or bitwise accurate emulation) wrt when processing simulation.

So, if you confirm I can adapt my downstream code assuming the encoding will eventually become `F = (J+0.5)*lsb ; J = floor(F/lsb)` in here.

BTW, if two-complement is now used for integers, wouldn't it make more sense to return the signed ones as `ap_int` instead of `ap_uint` in the accessors? 

Giovanni

Reply:
upsi I messed just up; of course I call `J = floor(F / lsb); F = (J + .5) * lsb` biased, sorry for that.